lib/opte/src/api.rs

```diff
@@ -0,0 +1,156 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2025 Oxide Computer Company
+
+pub use opte_api::*;
+
+use core::fmt::Display;
+use core::fmt::{self};
+#[cfg(feature = "engine")]
+use core::hash::Hash;
+#[cfg(feature = "engine")]
+use crc32fast::Hasher;
+use serde::Deserialize;
+use serde::Serialize;
+
+const AF_INET: i32 = 2;
+const AF_INET6: i32 = 26;
+
+pub static FLOW_ID_DEFAULT: InnerFlowId = InnerFlowId {
+    proto: 255,
+    addrs: AddrPair::V4 { src: Ipv4Addr::ANY_ADDR, dst: Ipv4Addr::ANY_ADDR },
+    src_port: 0,
+    dst_port: 0,
+};
+
+/// The flow identifier.
+///
+/// In this case the flow identifier is the 5-tuple of the inner IP
+/// packet.
+///
+/// NOTE: This should not be defined in `opte`. Rather, the engine
+/// should be generic in regards to the flow identifier, and it should
+/// be up to the `NetworkImpl` to define it.
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Deserialize,
+    Eq,
+    Hash,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Serialize,
+)]
+#[repr(C, align(4))]
+pub struct InnerFlowId {
+    // Using a `u8` here for `proto` hides the enum repr from SDTs.
+    pub proto: u8,
+    // We could also theoretically get to a 38B packing if we reduce
+    // AddrPair's repr from `u16` to `u8`. However, on the dtrace/illumos
+    // side `union addrs` is 4B aligned -- in6_addr_t has a 4B alignment.
+    // So, this layout has to match that constraint -- placing addrs at
+    // offset 0x2 with `u16` discriminant sets up 4B alignment for the
+    // enum variant data (and this struct itself is 4B aligned).
+    pub addrs: AddrPair,
+    pub src_port: u16,
+    pub dst_port: u16,
+}
+
+impl InnerFlowId {
+    #[cfg(feature = "engine")]
+    pub fn crc32(&self) -> u32 {
+        let mut hasher = Hasher::new();
+        self.hash(&mut hasher);
+        hasher.finalize()
+    }
+}
```

Comment:
The main reason this type lives here and not in `opte_api` is so that we can have this `crc32` method implemented directly on the flow ID, while keeping `crc32fast` limited to `engine`. Adding a conditional import to `opte_api` itself felt... bad.

In the fullness of time (#754), we probably want our base `FlowId` trait, at which point `opte::engine` can add its own `FlowIdExt: FlowId` trait with methods like this.